Schemas.md

```diff
@@ -0,0 +1,521 @@
+        * [Case 1: object_class key](#case-1-object_class-key)
+        * [Case 2: validation_method key](#case-2-validation_method-key)
+      * [Extra validators](#extra-validators)
+      * [Extra arguments](#extra-arguments)
+      * [Handlers with no arguments](#handlers-with-no-arguments)
+      * [Post schema operations](#post-schema-operations)
+    * [Common errors faced](#common-errors-faced)
+    * [Examples for reference](#examples-for-reference)
+    * [Debugging tricks](#debugging-tricks)
+    * [Contact](#contact)
+
+## Introduction
+
+At Oppia, we often want to describe the possible structure of some data. This data could be as simple as the boolean value of a checkbox in a user's preferences, or it could be more complex like the valid arguments to one of our backend API controllers. We use describe these data using _schemas_.
+
+Schemas describe the structure of data in a machine-readable format so that generic utility functions can operate on arbitrary data so long as the data comes with a schema. For example, schemas let us:
+
+* Validate provided data to make sure it conforms to the expected schema.
+* Automatically generate forms that let a user provided data in a form specified by a schema.
+
+## Write a schema
+
+A schema is just a dictionary that takes the form described below:
+
+* Required keys:
+
+  * `type`: The type of the data. Must be one of the following types. Unless otherwise stated, the type corresponds to the Python type of the same name.
+
+    * `bool`
+    * `int`
+    * `float`
+    * `basestring`: A type that includes both the `bytes` and `str` types in Python 3. Normalization will _not_ decode the data, even if a `bytes` object is provided.
+    * `unicode`: The `str` type in Python 3, which was the `unicode` type in Python 2. This type also accepts a `bytes` object, which will be automatically decoded according to UTF-8 during normalization.
+    * `list`
+    * `dict`
+    * `html`: A string (`bytes` or `str` in Python 3) with HTML code. When normalized, the HTML code will be sanitized by `core.domain.html_cleaner.clean()`. A `bytes` object may also be provided, which will be automatically decoded acording to UTF-8 during normalization.
+    * `custom`: A custom object type defined in `extensions/objects/models/objects.py`.
+    * `object_dict`: A dictionary that corresponds to a domain object (an object whose class is defined in the domain layer).
+    * `unicode_or_none`: Accepts data of type `unicode` or `None`. If the data is not `None`, it is normalized as if it were of type `unicode`.
+
+* Optional keys:
+
+  * `choices`: A list of possible values. The data must exactly match one of the choices.
+  * `validators`: A list of dictionaries, each of which has a sole key `id` that maps to the name of a validation function. The validation functions are defined as static methods of `schema_utils._Validators`. To conform to the schema, data must pass all validators.
```

Comment:
...all the validators specified in the schema.

Reply:
done